packages/xo-server-sdn-controller/src/ovsdb-client.js

```
@@ -0,0 +1,521 @@
+import assert from 'assert'
+import createLogger from '@xen-orchestra/log'
+import forOwn from 'lodash/forOwn'
+import fromEvent from 'promise-toolbox/fromEvent'
+import { connect } from 'tls'
+
+const log = createLogger('xo:xo-server:sdn-controller:ovsdb-client')
+
+const OVSDB_PORT = 6640
+
+// =============================================================================
+
+export class OvsdbClient {
+  constructor(host, clientKey, clientCert, caCert) {
+    this._host = host
+    this._numberOfPortAndInterface = 0
+    this._requestID = 0
+
+    this.updateCertificates(clientKey, clientCert, caCert)
+
+    log.debug(`[${this._host.name_label}] New OVSDB client`)
+  }
+
+  // ---------------------------------------------------------------------------
+
+  get address() {
+    return this._host.address
+  }
+
+  get host() {
+    return this._host.$ref
+  }
+
+  get id() {
+    return this._host.$id
+  }
+
+  updateCertificates(clientKey, clientCert, caCert) {
+    this._clientKey = clientKey
+    this._clientCert = clientCert
+    this._caCert = caCert
+
+    log.debug(`[${this._host.name_label}] Certificates have been updated`)
+  }
+
+  // ---------------------------------------------------------------------------
+
+  async addInterfaceAndPort(
+    networkUuid,
+    networkName,
+    remoteAddress,
+    encapsulation
+  ) {
+    const socket = await this._connect()
+    if (socket == null) {
+      log.error(`[${this._host.name_label}] No TLS socket available`)
+      return
```

I'm not sure it's a good idea to hide the error from the calling code.